src/doc/model.lisp

```diff
@@ -0,0 +1,303 @@
+         (env entry:*global-environment*)
+         (type (tc:lookup-value-type env name))
+         (source-param-names (tc:lookup-function-source-parameter-names env name)))
+    (if source-param-names
+        (make-instance 'function-entry
+          :name name
+          :type type
+          :documentation documentation
+          :location location
+          :param-names source-param-names)
+        (make-instance 'value-entry
+          :name name
+          :type type
+          :documentation documentation
+          :location location))))
+
+(defun make-class-entry (entry &key package)
+  (make-instance 'class-entry
+    :name (tc:ty-class-name entry)
+    :context (tc:ty-class-superclasses entry)
+    :predicate (tc:ty-class-predicate entry)
+    :methods (remove-if (lambda (method)
+                          (and package (exported-symbol-p (tc:ty-class-method-name method) package t)))
+                        (tc:ty-class-unqualified-methods entry))
+    :instances (reverse (fset:convert 'list (tc:lookup-class-instances entry:*global-environment* (tc:ty-class-name entry) :no-error t)))
+    :documentation (tc:ty-class-docstring entry)
+    :location (tc:ty-class-location entry)))
+
+(defun make-type-entry (entry ctors applicable-instances)
+  (let* ((name (tc:type-entry-name entry))
+         (struct-entry (tc:lookup-struct entry:*global-environment* name :no-error t)))
+    (cond ((not (null struct-entry))
+           (make-instance 'struct-entry
+             :name name
+             :documentation (tc:type-entry-docstring entry)
+             :location (tc:type-entry-location entry)
+             :type (tc:type-entry-type entry)
+             :tyvars (tc:type-entry-tyvars entry)
+             :fields (tc:struct-entry-fields struct-entry)
+             :instances applicable-instances))
+          (t
+           (make-instance 'type-entry
+             :name name
+             :type (tc:type-entry-type entry)
+             :constructors ctors
+             :constructor-types (mapcar
+                                 (lambda (ctor)
+                                   (tc:lookup-value-type entry:*global-environment*
+                                                         (tc:constructor-entry-name ctor)))
+                                 ctors)
+             :instances applicable-instances
+             :documentation (tc:type-entry-docstring entry)
+             :location (tc:type-entry-location entry))))))
+
+;; helper
+
+(defun applicable-p (type-entry instance)
+  (some (lambda (pred-type)
+          (labels ((check (pred)
+                     (typecase pred
+                       (tc:tapp (check (tc:tapp-from pred)))
+                       (t (equalp (tc:type-entry-type type-entry) pred)))))
+            (check pred-type)))
+        (tc:ty-predicate-types (tc:ty-class-instance-predicate instance))))
+
+(defun constructs-p (type-entry constructor-entry)
+  (eql (tc:constructor-entry-constructs constructor-entry)
```

Comment:
check if eql is correct?

(if symbols use `eq`)

Reply:
it's eq